### PR TITLE
merge Reshare and Resharable traits

### PR DESCRIPTION
**Pull request by andyleiserson**
Also change `reshare` to be a method (take `&self`).

### CODE REVIEW

File: src/protocol/basics/reshare.rs [hunk 1]
```diff
@@ -95,13 +37,28 @@ impl<F: Field> LegacyReshare<F> for MaliciousContext<'_, F> {
...
+        to_helper: Role,
+    ) -> Result<Self, Error>
+    where
+        C: 'fut;
+}
+
+// Variant of Reshare implemented for types like [T] where resharing doesn't return `Self`.  This is
+// just a more general version of `Reshare`, but a bunch of consumers that don't otherwise care
+// would need to specify an associated type bound `T: Reshare<Output = T>` to use this more general
+// version.
```
Comment (martinthomson) [inline src/protocol/basics/reshare.rs:54]: 👍  If only there was a way to provide a default for an associated type... https://github.com/rust-lang/rust/issues/29661

**Review by martinthomson — APPROVED** [review]
Goodness continues.

**Comment by taikiy** [comment]
❤❤❤

**Comment by akoshelev** [comment]
merging it as we got approvals

**Pull request merged**